azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py

```diff
@@ -137,60 +121,89 @@ def _on_transition_complete(event_data):
         """
+        logger.info("Calling provider connect")
         self._mqtt_provider.connect()
 
-    def _after_action_provider_disconnect(self, event_data):
+    def _call_provider_disconnect(self, event_data):
         """
         Call into the provider to disconnect the transport.
-        This is meant to be called by the state machine as an "after" action
+        This is meant to be called by the state machine as part of a state transition
         """
+        logger.info("Calling provider disconnect")
         self._mqtt_provider.disconnect()
 
-    def _before_action_add_event_to_queue(self, event_data):
+    def _on_provider_connect_complete(self):
         """
-        Queue an event for sending later.
-        This is meant to be called by the state machine as an "before" action
+        Callback that is called by the provider when the connection has been established
         """
-        # TODO: throw here if event_data.args[0] is not an event/message
-        self._event_queue.put_nowait(event_data.args[0])
+        logger.info("_on_provider_connect_complete")
+        self._trig_provider_connect_complete()
+
+        if self.on_transport_connected:
+            self.on_transport_connected("connected")
+        callback = self._connect_callback
+        if callback:
+            self._connect_callback = None
+            callback()
 
-    def _queue_is_empty(self, event_data):
+    def _on_provider_disconnect_complete(self):
+        """
+        Callback that is called by the provider when the connection has been disconnected
         """
-        Return true if the sending queue is empty.
-        This is meant to be called by the state machine as a "conditions" or "unless" check
+        logger.info("_on_provider_disconnect_complete")
+        self._trig_provider_disconnect_complete()
+
+        if self.on_transport_disconnected:
+            self.on_transport_disconnected("disconnected")
+        callback = self._disconnect_callback
+        if callback:
+            self._disconnect_callback = None
+            callback()
+
+    def _on_provider_publish_complete(self, mid):
         """
-        return self._event_queue.empty()
+        Callback that is called by the provider when a publish operation is complete.
+        """
+        if self.on_event_sent:
+            self.on_event_sent()
+        if mid in self._event_callback_map:
+            callback = self._event_callback_map[mid]
+            del self._event_callback_map[mid]
+            callback()
+
+    def _add_event_to_queue(self, event_data):
+        """
+        Queue an event for sending later.  All events that get sent end up going into
+        this queue, even if they're going to be sent immediately.
+        """
+        logger.info("Adding event to queue for later sending")
+        # TODO: throw here if event_data.args[0] is not an event/message
+        self._event_queue.put_nowait((event_data.args[0], event_data.args[1]))
```

Comment:
>((event_data.args[0], event_data.args[1])) [](start = 36, length = 42)

args in arrays are going to be hard to track, nothing we can do about this?

Reply:
this comes from transitions.  They pack this for us.  I don't like this either. 

---
In reply to: [237662131](https://github.com/Azure/azure-iot-sdk-python-preview/pull/26#discussion_r237662131) [](ancestors = 237662131)